asim: measure thrashing

wenyihu6 opened the pull request:
This commit adds a thrashing benchmark to each stat output over ticks. Thrashing
is measured by frequency and magnitude: frequency is the number of sign changes
in the first derivative (slope), and magnitude is the absolute difference at
each sign change, normalized by the mean value across ticks. The final scalar
thrashing score is the maximum thrashing value observed across all stores.

Epic: none 
Release note: noe

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/151341)
<!-- Reviewable:end -->

tbg reviewed (changes requested):
Thank you for doing this! I had a few small suggestions.

> @tbg reviewed 26 of 26 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151341#-:-OYWH-1hExHq-ejumh25:b-vqg864)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[`pkg/kv/kvserver/asim/history/history.go` line 56 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/151341#-OYWFAsmDAa_3XRpilPq:-OYWFAsmDAa_3XRpilPr:b-tgcjp0) ([raw file](https://github.com/cockroachdb/cockroach/blob/715a1137cdca6aa05866285ad24725b39d20adde/pkg/kv/kvserver/asim/history/history.go#L56)):*
> ```Go
> // the sum of the absolute differences between the values when the sign of the
> // derivative changes divided by mean value over all ticks.
> type volatility struct {
> ```

nit: maybe `thrashing` is a better name here. "Volatility" suggests that we're measuring something similar to quadratic variation (i.e. that we're riffing off the definition of volatility in finance), but we're not.
___
*[`pkg/kv/kvserver/asim/history/history.go` line 57 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/151341#-OYW6aYUCJKNBFihg-tp:-OYW6aYUCJKNBFihg-tq:b9c05j5) ([raw file](https://github.com/cockroachdb/cockroach/blob/715a1137cdca6aa05866285ad24725b39d20adde/pkg/kv/kvserver/asim/history/history.go#L57)):*
> ```Go
> // derivative changes divided by mean value over all ticks.
> type volatility struct {
> 	frequency int
> ```

frequency should be in units 1/s, so it would need to be the count divided by the time duration (represented by the number of ticks). Maybe this should just be a count (i.e. the name reflecting that).
___
*[`pkg/kv/kvserver/asim/history/history.go` line 62 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/151341#-OYW6jsH1tX8RbW9YFmH:-OYW6jsH1tX8RbW9YFmI:b-mt06bq) ([raw file](https://github.com/cockroachdb/cockroach/blob/715a1137cdca6aa05866285ad24725b39d20adde/pkg/kv/kvserver/asim/history/history.go#L62)):*
> ```Go
> 	// changes. But it is possible that the magnitude is small at that time but
> 	// change aggressively soonly after.
> 	magnitude float64
> ```

How about using the total variation over the turning points (points where the delta between neighboring values flips) here: if the turning points are `x_0, x_1, ..., x_n` sum up `|f(x_{k}) - f(x_{k-1})|` (where `1 <= k <= n`).
For example, for a sin wave `sin(t)` this picks up `2` for each time period of `pi` (starting at `k*pi + pi/2`). For monotonic functions, it picks up nothing. For a function that is monotonic at first but begins to jitter only at the end, it will only pick up on the non-monotonic portions.

As for normalization, the mean isn't a bad start to make values comparable, but I think the `max-min` could be better - it's the effective "spread" of the entire data, which I think is what we want to relate the fluctuations to. (For example, we could say "in this run, the replica count fluctuated by 17% (of the observed range, which may have been something like 1200 replicas at the peak and 300 replicas at the minimum).
It would be somewhat better  - I think - to normalize by something that is consistent across runs, i.e. that is a property of the input data and not the output. I'd kick that can down the road, though. We could try to normalize by the max(across all stores) for the first datapoint, which represents the initial conditions; that should work well for at least some metrics.

I also think there's use for the un-normalized version to be kept around, since that directly corresponds to work done by the system at least for some metrics. If replica count oscillates between 100 and 200, we moved 100 replicas around, period.

After all this musing, my suggestion would be:

- switch to total variation within turning point regions
- normalize by `max-min`
- also add the un-normalized reading to the struct


<!-- Sent from Reviewable.io -->

tbg committed:
asim: measure thrashing

This commit adds a thrashing benchmark to each stat output. "Thrashing" can be
measured in many ways, but here we measure, for each per-store timeseries, the
total variation (but excluding the first monotonic run). When displaying the
resulting value, we additionally normalize (divide by) the "range" of values
seen (i.e. the max minus the min), and print the result as a percentage. For
example, the time series [0, 1, 0, 1, 0, 1] would be assigned a thrashing
measurement of 200%, because:

- the range is 1 (two values are at most 1 apart from each other)
- there are three runs, each [0, 1], but we're ignoring the first one
- the remaining two runs contribute total variation of 1, each
- (1+1)/1 = 2 = 200%.

Intuitively, this means that the time series "sweeps" across the range
twice. Additional oscillations would drive this number up further.

When printing the thrashing across a set of stores (but for a fixed stat),
instead of computing the range within each store's time series, we compute it
across the entire set. This makes it more likely that we discover the "true
set".

All of this can be made arbitrarily more complicated (and additional
normalization by the number of runs or ticks may be appropriate for some use
cases), but this should be a good start.

Epic: CRDB-49117
Release note: none

tbg committed:
history: add more thrashing test cases

tbg committed:
asim: improve thrashing measurement

The previous mechanism was "total variation in all but the first run", but this
didn't do well on examples in which the series was basically monotonic but with
an early outlier (in which case the bulk of the "good monotonic function" would
also be tallied up as thrashing).

This commit introduces and adopts a "trend-discounting total variation".
Basically, it looks at the relative size of the upward and downward variation
and, if one dominates, discounts it in the computed value.

This should provide much better results. For example, in one instance from
example_liveness, we see the computed thrashing measure drop from 145% to 6%,
reflecting that most replica movements "went in the correct direction", which
wasn't previously captured properly.

tbg commented:
Apologies for the accidental force push to this PR (I named my branch the same by accident 😓 )
Closing in favor of https://github.com/cockroachdb/cockroach/pull/152506.
